sdk/core/Azure.Core/src/Shared/SyncAsyncEventHandlerExtensions.cs

```diff
@@ -0,0 +1,123 @@
+        /// </returns>
+        /// <exception cref="AggregateException">
+        /// An exception was thrown during the execution of at least one of the
+        /// event's handlers.
+        /// </exception>
+        /// <exception cref="ArgumentNullException">
+        /// Thrown when <paramref name="e"/>, <paramref name="declaringTypeName"/>,
+        /// <paramref name="eventName"/>, or <paramref name="clientDiagnostics"/>
+        /// are null.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// Thrown when <paramref name="declaringTypeName"/> or
+        /// <paramref name="eventName"/> are empty.
+        /// </exception>
+        public static async Task RaiseAsync<T>(
+            this SyncAsyncEventHandler<T> eventHandler,
+            T e,
+            string declaringTypeName,
+            string eventName,
+            ClientDiagnostics clientDiagnostics)
+            where T : SyncAsyncEventArgs
+        {
+            Argument.AssertNotNull(e, nameof(e));
+            Argument.AssertNotNullOrEmpty(declaringTypeName, nameof(declaringTypeName));
+            Argument.AssertNotNullOrEmpty(eventName, nameof(eventName));
+            Argument.AssertNotNull(clientDiagnostics, nameof(clientDiagnostics));
+
+            // Get the invocation list, but return early if there's no work
+            if (eventHandler == null) { return; }
+            Delegate[] handlers = eventHandler.GetInvocationList();
+            if (handlers == null || handlers.Length == 0) { return; }
+
+            // Wrap handler invocation in a distributed tracing span so it's
+            // easy for customers to track and measure
+            string eventFullName = declaringTypeName + "." + eventName;
+            using DiagnosticScope scope = clientDiagnostics.CreateScope(eventFullName);
+            scope.Start();
+            try
+            {
+                // Collect any exceptions raised by handlers
+                List<Exception> failures = null;
+
+                // Raise the handlers sequentially so we don't introduce any
+                // unintentional parallelism in customer code
+                foreach (Delegate handler in handlers)
+                {
+                    SyncAsyncEventHandler<T> azureHandler = (SyncAsyncEventHandler<T>)handler;
+                    try
+                    {
+                        Task runHandlerTask = azureHandler(e);
+                        // We can consider logging something when e.RunSynchronously
+                        // is true, but runHandlerTask.IsComplete is false.
+                        // (We're not bother to check our tests because
+                        // EnsureCompleted on the code path that raised the
+                        // event will catch it for us.)
+                        await runHandlerTask.ConfigureAwait(false);
+                    }
+                    catch (Exception ex)
```

Comment:
No special handling for `OperationCanceledException`?

Reply:
No, see the discussion around Heath's comment.